### ISSUE TITLE
rotate 180 degrees unexpectedly adds a displacement

### ISSUE THREAD

**Issue by ewelot** [libvips/libvips#3271]
When applying image rotation by 180 degrees there seems to be a horizontal and vertical displacement added.

Commands to reproduce:
```
convert rose: rose.jpg
vips rotate rose.jpg rose_rot.jpg 180
```
As a result pixels are shifted to the bottom right and a black border appears at top row and left column.
I am running on libvips version 8.13.3.

**Comment by jcupitt** [comment]
Yes, there's an issue for this somewhere, affine still (!!) has a couple of corner cases like this.

As a workaround, use `vips_rot()` for exact rotations.

**Comment by ewelot** [comment]
Back on this issue again as I need to rely on rotate() for using arbitrary rotation angles ...
1. Could you please elaborate on the meaning of the optional offset parameters?
2. In the simple case of not using any of those offset parameters I would assume a rotation about the image center. As a result the output image width should be an even number if input image width is even and and odd number if input image width is odd. The same applies for image heights. Unfortunately this is not true, see e.g.
```
convert rose: rose.jpg
vips rotate rose.jpg rose_rot.jpg 40
identify rose.jpg rose_rot.jpg
```
which outputs
```
rose.jpg JPEG 70x46 70x46+0+0 8-bit sRGB 2664B 0.000u 0:00.000
rose_rot.jpg JPEG 83x80 83x80+0+0 8-bit sRGB 2120B 0.000u 0:00.000

```
Does this indicate a problem with the coordinates of the center of rotation (e.g. the difference between the image pixel "index" coordinate and a cartesian system having its origin at pixel edges)?

**Comment by jcupitt** [comment]
Hello again,

I think it's corner vs. centre sampling. Imagine a coordinate grid (sorry for the ascii art):

```
A---+---+---+- ...
|   |   |   |
+---B---+---+-
|   |   |   |
+---+---+---+-
|   |   |   |
.
.
```

where `A` is (0, 0), `B` is (1, 1), etc.

Now imagine a matrix of numbers:

```
1 2 3 ..
4 5 6
. 
.
```

What's the value of the point (0, 0) on the grid? libvips makes it 1 by default, which is corner sampling. This works well for resizing, but you shift by 1 pixel when you rotate 90 degrees.

The alternative convention people commonly use is centre sampling. Now you put the 1 at the centre of the top-left pixel, ie. at (0.5, 0.5). Now when you rotate, there's no mysterious extra shift, but there's an extra 0.5 to take into account for scaling.

The `idx` and `idy` parameters in `affine` set an offset in the input space, so setting them to (-0.5, -0.5) turns the libvips corner sample default into centre sampling.

**Comment by ewelot** [comment]
Hi John,

I like your ascii art so much that I am doing it that way too ... ;)

Yes, its a matter of center or edge/corner sampling.
But in case of not using any offsets I would assume the true image center is
preserved regardless if it is located on the center/corner/edge of a pixel. The
center of rotation (C) should probably be as in the following images

+---+---+---+    +---+---+---+---+    +---+---+---+     +---+---+---+---+
|   |   |   |    |   |   |   |   |    |   |   |   |     |   |   |   |   |
+---+---+---+    +---+---+---+---+    +---+---+---+     +---+---+---+---+
|   | C |   |    |   |   C   |   |    |   |   |   |     |   |   |   |   |
+---+---+---+    +---+---+---+---+    +---+-C-+---+     +---+---C---+---+
|   |   |   |    |   |   |   |   |    |   |   |   |     |   |   |   |   |
+---+---+---+    +---+---+---+---+    +---+---+---+     +---+---+---+---+
                                      |   |   |   |     |   |   |   |   |
                                      +---+---+---+     +---+---+---+---+

and output image width/height should have numbers which are
a) odd/odd       b) even/odd          c) odd/even       d) even/even

Doesn't this look like what it is expected for most people?

Regards
Thomas


> Am Wed, 18 Sep 2024 04:30:28 -0700
> schrieb John Cupitt ***@***.***>:
>
> Hello again,
> 
> I think it's corner vs. centre sampling. Imagine a coordinate grid (sorry for the ascii art):
> 
> ```
> A---+---+---+- ...
> |   |   |   |
> +---B---+---+-
> |   |   |   |
> +---+---+---+-
> |   |   |   |
> .
> .
> ```
> 
> where `A` is (0, 0), `B` is (1, 1), etc.
> 
> Now imagine a matrix of numbers:
> 
> ```
> 1 2 3 ..
> 4 5 6
> . 
> .
> ```
> 
> What's the value of the point (0, 0) on the grid? libvips makes it 1 by default, which is corner sampling. This works well for resizing, but you shift by 1 pixel when you rotate 90 degrees.
> 
> The alternative convention people commonly use is centre sampling. Now you put the 1 at the centre of the top-left pixel, ie. at (0.5, 0.5). Now when you rotate, there's no mysterious extra shift, but there's an extra 0.5 to take into account for scaling.
> 
> The `idx` and `idy` parameters in `affine` set an offset in the input space, so setting them to (-0.5, -0.5) turns the libvips corner sample default into centre sampling.
>
